File: C++/Laba3sem2v2/stack.cpp
#include "stack.h"
#include "array.h"

#include <cstring>
#include <fstream>
#include <iostream>

using namespace std;

// Инициализация стека.
void Stack::init() {
    top = nullptr;
}

// Добавление элемента на вершину стека.
void Stack::push(const string& value) {
    StackNode* newNode = new StackNode{value, top};
    top = newNode;
}

// Удаление элемента с вершины стека.
void Stack::pop() {
    if (top == nullptr) {
        return;
    }
    StackNode* temp = top;
    top = top->next;
    delete temp;
}

// Вывод содержимого стека в консоль.
void Stack::print() {
    StackNode* temp = top;
    while (temp != nullptr) {
        cout << temp->data << " ";
        temp = temp->next;
    }
    cout << endl;
}

// Освобождение памяти, занимаемой стеком.
void Stack::destroy() {
    while (top != nullptr) {
        pop();
    }
}

// Загрузка стека из файла с использованием DynamicArray.
void Stack::loadFromFile(const string& fileName) {
    ifstream file(fileName);
    if (!file.is_open()) {
        return;
    }

    // Создаем временный динамический массив для хранения строк из файла.
    DynamicArray tempArray;

    string value;
    // Считываем все строки из файла в наш временный массив.
    while (file >> value) {
        tempArray.add(value);
    }
    file.close();

    // Теперь проходим по временному массиву С КОНЦА В НАЧАЛО,
    // чтобы правильно воссоздать стек.
    for (int i = tempArray.length() - 1; i >= 0; --i) {
        push(tempArray.get(i));
    }
}

// Сохранение стека в файл. Эта функция остается простой.
void Stack::saveToFile(const string& fileName) {
    ofstream file(fileName);
    StackNode* temp = top;
    // Просто проходим по стеку от вершины к основанию и записываем в файл.
    while (temp != nullptr) {
        file << temp->data << endl;
        temp = temp->next;
    }
    file.close();
}


// Функция для управления стеком через командную строку (остается без изменений).
void runStack(int argc, char* argv[]) {
    Stack stack;
    stack.init();

    string fileName;
    string query;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--file") == 0 && i + 1 < argc) {
            fileName = argv[i + 1];
            i++;
        } else if (strcmp(argv[i], "--query") == 0 && i + 1 < argc) {
            query = argv[i + 1];
            i++;
        }
    }

    if (!fileName.empty()) {
        stack.loadFromFile(fileName);
    }

    string command;
    size_t pos = query.find(' ');
    if (pos != string::npos) {
        command = query.substr(0, pos);
        query = query.substr(pos + 1);
    } else {
        command = query;
    }

    if (command == "SPUSH") {
        stack.push(query);
        if (!fileName.empty()) {
            stack.saveToFile(fileName);
        }
    } else if (command == "SPOP") {
        stack.pop();
        if (!fileName.empty()) {
            stack.saveToFile(fileName);
        }
    } else if (command == "SPRINT") {
        stack.print();
    }

    stack.destroy();
}

File: sem3l3go/stack/stack.go
package stack

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

// StackNode представляет узел стека
type StackNode struct {
	Data string
	Next *StackNode
}

// Stack представляет стек
type Stack struct {
	Top *StackNode
}

// NewStack создает новый стек
func NewStack() *Stack {
	return &Stack{
		Top: nil,
	}
}

// Push добавляет элемент на вершину
func (s *Stack) Push(value string) {
	s.Top = &StackNode{Data: value, Next: s.Top}
}

// Pop удаляет элемент с вершины
func (s *Stack) Pop() bool {
	if s.Top == nil {
		return false
	}
	s.Top = s.Top.Next
	return true
}

// Print выводит все элементы
func (s *Stack) Print() {
	temp := s.Top
	for temp != nil {
		fmt.Printf("%s ", temp.Data)
		temp = temp.Next
	}
	fmt.Println()
}

// Destroy очищает стек
func (s *Stack) Destroy() {
	for s.Top != nil {
		s.Pop()
	}
}

// LoadFromFile загружает из файла
func (s *Stack) LoadFromFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Open(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	var values []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		values = append(values, scanner.Text())
	}

	// Загружаем в обратном порядке для правильного порядка стека
	for i := len(values) - 1; i >= 0; i-- {
		s.Push(values[i])
	}
	return scanner.Err()
}

// SaveToFile сохраняет в файл
func (s *Stack) SaveToFile(filename string) error {
	if filename == "" {
		return nil
	}
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	temp := s.Top
	for temp != nil {
		if _, err := fmt.Fprintln(file, temp.Data); err != nil {
			return err
		}
		temp = temp.Next
	}
	return nil
}

// RunStack выполняет команды над стеком
func RunStack(args []string) {
	stack := NewStack()
	var filename string
	var query string

	for i := 0; i < len(args); i++ {
		if args[i] == "--file" && i+1 < len(args) {
			filename = args[i+1]
			i++
		} else if args[i] == "--query" && i+1 < len(args) {
			query = args[i+1]
			i++
		}
	}

	if filename != "" {
		stack.LoadFromFile(filename)
	}

	parts := strings.SplitN(query, " ", 2)
	command := parts[0]
	arg := ""
	if len(parts) > 1 {
		arg = parts[1]
	}

	switch command {
	case "SPUSH":
		if arg != "" {
			stack.Push(arg)
			if filename != "" {
				stack.SaveToFile(filename)
			}
		}
	case "SPOP":
		stack.Pop()
		if filename != "" {
			stack.SaveToFile(filename)
		}
	case "SPRINT":
		stack.Print()
	}

	stack.Destroy()
}
